### PR TITLE
filters: optimize cel expression context with constant-time lookups

### PR DESCRIPTION
**Pull request by agrawroh**
## Description

This PR optimizes the CEL expression context implementation by replacing `if-else` chains with hash maps for `O(1)` lookups and improving memory management. This would enhance the performance for header and filter state access patterns while maintaining the same functionality.

This change introduces constant-time lookups using `absl::flat_hash_map` and improves the overall design through static initialization of lookup maps for better performance.

### Before

```
Run on (32 X 3364.47 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x16)
  L1 Instruction 32 KiB (x16)
  L2 Unified 2048 KiB (x16)
  L3 Unified 107520 KiB (x1)
Load Average: 9.07, 16.88, 15.22
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
bmRequestAttributes/100             0.179 us        0.179 us      3943200
bmRequestAttributes/10000           0.178 us        0.178 us      3768352
bmRequestAttributes/1000000         0.179 us        0.179 us      3882288
bmResponseAttributes/100            0.187 us        0.187 us      3734635
bmResponseAttributes/10000          0.180 us        0.180 us      3778717
bmResponseAttributes/1000000        0.196 us        0.196 us      3668511
bmFilterState/1000                   51.6 us         51.6 us        10000
bmFilterState/10000                   692 us          692 us         1025
bmFilterState/1000000              262413 us       262406 us            3
bmFilterState/10000000            3887577 us      3887332 us            1
bmConnectionAttributes/100          0.465 us        0.465 us      1521335
bmConnectionAttributes/10000        0.467 us        0.467 us      1500911
bmConnectionAttributes/1000000      0.468 us        0.468 us      1521145
```

### After

```
Run on (32 X 3269.07 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x16)
  L1 Instruction 32 KiB (x16)
  L2 Unified 2048 KiB (x16)
  L3 Unified 107520 KiB (x1)
Load Average: 15.79, 17.71, 15.83
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
bmRequestAttributes/100             0.166 us        0.166 us      4024447
bmRequestAttributes/10000           0.167 us        0.167 us      4027806
bmRequestAttributes/1000000         0.171 us        0.171 us      4217833
bmResponseAttributes/100            0.174 us        0.174 us      3933317
bmResponseAttributes/10000          0.169 us        0.169 us      4106892
bmResponseAttributes/1000000        0.173 us        0.173 us      4110847
bmFilterState/1000                   52.1 us         52.1 us        14034
bmFilterState/10000                   655 us          655 us         1078
bmFilterState/1000000              251507 us       251496 us            3
bmFilterState/10000000            3806409 us      3806192 us            1
bmConnectionAttributes/100          0.435 us        0.435 us      1613013
bmConnectionAttributes/10000        0.423 us        0.423 us      1674895
bmConnectionAttributes/1000000      0.411 us        0.411 us      1729797
```

---

**Commit Message:** filters: optimize cel expression context with constant-time lookups
**Additional Description:** This commit replaces linear if-else chains with flat_hash_maps for property lookups in the CEL expression context. It also improves memory management and provides better encapsulation of implementation details.

**Risk Level:** Low
- No functional changes
- Only performance optimizations
- Existing test coverage maintained

**Testing:**
- All existing tests pass
- No behavioral changes

**Docs Changes:** N/A - internal implementation change only
**Release Notes:** N/A

### CODE REVIEW

**Comment by wbpcode** [comment]
Thanks. This is great. Any performance improvements are welcome if it does't break the security and our style.

But could you also add a benchmark test to prove it works as our expection? Thanks.

/wait

**Comment by agrawroh** [comment]
> Thanks. This is great. Any performance improvements are welcome if it does't break the security and our style.
> 
> 
> 
> But could you also add a benchmark test to prove it works as our expection? Thanks.
> 
> 
> 
> /wait

@wbpcode Thanks for taking a look. I can certainly add a benchmarking test. This would be my first time. If possible, could you please share some pointers on how to write such a test?

**Comment by wbpcode** [comment]
> > Thanks. This is great. Any performance improvements are welcome if it does't break the security and our style.
> > But could you also add a benchmark test to prove it works as our expection? Thanks.
> > /wait
> 
> @wbpcode Thanks for taking a look. I can certainly add a benchmarking test. This would be my first time. If possible, could you please share some pointers on how to write such a test?

Every test is different. But you can find lots test files that end with `benchmark` or contain `speed`. There are lots of test cases. Besically, you only need to define a static function to run the code you want to test. And then the benchmark macro could help you to resolve other things.

**Comment by wbpcode** [comment]
Seems the performance improvement is trival. But I also agree the new look up table pattern is better. So, will take a look tomorrow.

**Comment by nezdolik** [comment]
cc @tyxia @kyessenov (extension owners)

File: source/extensions/filters/common/expr/context.cc [hunk 1]
```diff
@@ -39,242 +39,423 @@ convertHeaderEntry(Protobuf::Arena& arena,
 
 namespace {
 
+const absl::flat_hash_map<absl::string_view, SslExtractor>& getSslExtractors() {
+  static const auto* extractors = new absl::flat_hash_map<absl::string_view, SslExtractor>{
```
Comment (nezdolik) [inline source/extensions/filters/common/expr/context.cc:43]: prefer to use smart pointers instead (e.g. std::make_unique<absl::flat_hash_map<absl::string_view, SslExtractor>>() depending on the ownership model) here and in all `new` occurrences below

File: source/extensions/filters/common/expr/context.cc [hunk 2]
```diff
@@ -39,242 +39,423 @@ convertHeaderEntry(Protobuf::Arena& arena,
 
 namespace {
 
+const absl::flat_hash_map<absl::string_view, SslExtractor>& getSslExtractors() {
```
Comment (nezdolik) [inline source/extensions/filters/common/expr/context.cc:42]: nit: this can also be a static variable instead of a function

File: source/extensions/filters/common/expr/context.cc (hunk 1 above)
Comment (kyessenov) [reply source/extensions/filters/common/expr/context.cc:43]: You could use `ConstSingleton` with an object that has the map. This is a common way to store static tables.

File: source/extensions/filters/common/expr/context.cc [hunk 3]
```diff
@@ -334,14 +517,13 @@ absl::optional<CelValue> FilterStateWrapper::operator[](CelValue key) const {
     return {};
   }
   auto value = key.StringOrDie().value();
+
   if (const StreamInfo::FilterState::Object* object = filter_state_.getDataReadOnlyGeneric(value);
       object != nullptr) {
     const CelState* cel_state = dynamic_cast<const CelState*>(object);
     if (cel_state) {
       return cel_state->exprValue(&arena_, false);
-    } else if (object != nullptr) {
```
Comment (kyessenov) [inline source/extensions/filters/common/expr/context.cc:342]: This looks like unrelated change?

File: test/extensions/filters/common/expr/expr_context_speed_test.cc [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+#include "test/mocks/stream_info/mocks.h"
+#include "test/mocks/upstream/host.h"
+#include "test/test_common/utility.h"
+
+#include "benchmark/benchmark.h"
+#include "gmock/gmock.h"
+
+namespace Envoy {
+
+using testing::NiceMock;
```
Comment (kyessenov) [inline test/extensions/filters/common/expr/expr_context_speed_test.cc:18]: nit: this can be moved into the namespace of the extension.

File: test/extensions/filters/common/expr/expr_context_speed_test.cc [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        arena_, &request_headers_, info_);
+    response_ = std::make_unique<Extensions::Filters::Common::Expr::ResponseWrapper>(
+        arena_, &response_headers_, &response_trailers_, info_);
+    filter_state_ = std::make_unique<Extensions::Filters::Common::Expr::FilterStateWrapper>(
+        arena_, *info_.filterState());
+  }
+
+  void testRequestAttributes(::benchmark::State& state) {
+    static const std::vector<absl::string_view> attributes = {
+        "scheme", "path", "method", "referer", "user-agent", "size", "total_size", "protocol"};
```
Comment (kyessenov) [inline test/extensions/filters/common/expr/expr_context_speed_test.cc:50]: you can use the string_views for these attributes here

File: test/extensions/filters/common/expr/expr_context_speed_test.cc [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+namespace Envoy {
+
+using testing::NiceMock;
+using testing::Return;
+
+class ExpressionContextSpeedTest {
+public:
+  ExpressionContextSpeedTest(size_t num_attributes) {
+    ssl_info_ = std::make_shared<NiceMock<Ssl::MockConnectionInfo>>();
```
Comment (kyessenov) [inline test/extensions/filters/common/expr/expr_context_speed_test.cc:24]: nit: mocks are generally somewhat slow, so it's better to use real data structures for performance measurements.

**Comment by kyessenov** [comment]
LGTM apart from a change in filter_state lookup logic. I agree, this is better, and would allow us to implement `Keys()` quickly in a followup, even though the performance benefit is not significant.

File: source/extensions/filters/common/expr/context.cc (hunk 3 above)
Comment (agrawroh) [reply source/extensions/filters/common/expr/context.cc:342]: Oversight, undid it.

File: source/extensions/filters/common/expr/context.cc (hunk 1 above)
Comment (agrawroh) [reply source/extensions/filters/common/expr/context.cc:43]: I switched it to using `CONSTRUCT_ON_FIRST_USE`

File: test/extensions/filters/common/expr/expr_context_speed_test.cc (hunk 5 above)
Comment (agrawroh) [reply test/extensions/filters/common/expr/expr_context_speed_test.cc:50]: Done

File: source/extensions/filters/common/expr/context.h [hunk 7]
```diff
@@ -106,12 +106,65 @@ class WrapperFieldValues {
...
+
+class ResponseLookupValues {
+public:
+  const absl::flat_hash_map<absl::string_view, ResponseValueExtractor> response_lookup_;
+  static const ResponseLookupValues& get();
+};
+
+class ConnectionLookupValues {
+public:
+  const absl::flat_hash_map<absl::string_view, ConnectionValueExtractor> connection_lookup_;
```
Comment (tyxia) [inline source/extensions/filters/common/expr/context.h:141]: string_view doesn't own the data and could trigger lifetime issue like used in container case. I guess here is fine as long as you ensure the strings that are pointed to by string_view outlast the string_view itself.

But I am just curious any specific gain here by using string_view?

Comment (kyessenov) [reply source/extensions/filters/common/expr/context.h:141]: Memory savings? The key is not stored as a copy. I think this is safe since these are all constexpr keys.

Comment (agrawroh) [reply source/extensions/filters/common/expr/context.h:141]: @tyxia @kyessenov Is it okay to keep this as `string_view`?

**Review by tyxia — COMMENTED** [review]
LGTM (there is a open comment from Kuat regarding mocks in speed test)

Nice work,Thanks for contribution!

File: test/extensions/filters/common/expr/expr_context_speed_test.cc (hunk 6 above)
Comment (agrawroh) [reply test/extensions/filters/common/expr/expr_context_speed_test.cc:28]: @kyessenov I tried doing it but couldn't get it to work. The benchmark test shouldn't be blocking so if it's okay, I can try to take a look at it separately.

Comment (kyessenov) [reply test/extensions/filters/common/expr/expr_context_speed_test.cc:28]: Yes, it's fine since a map lookup is probably roughly mock virtual function table lookup, so results are not dominated by mocks overhead.

File: source/extensions/filters/common/expr/context.cc [hunk 8]
```diff
@@ -29,119 +29,450 @@ convertHeaderEntry(Protobuf::Arena& arena,
...
+          {DNSSanPeerCertificate,
+           [](const Ssl::ConnectionInfo& info) {
+             return !info.dnsSansPeerCertificate().empty()
+                        ? CelValue::CreateString(&info.dnsSansPeerCertificate()[0])
+                        : CelValue::CreateNull();
+           }},
+          {SHA256PeerCertificateDigest, [](const Ssl::ConnectionInfo& info) {
+             return !info.sha256PeerCertificateDigest().empty()
+                        ? CelValue::CreateString(&info.sha256PeerCertificateDigest())
+                        : CelValue::CreateNull();
```
Comment (kyessenov) [inline source/extensions/filters/common/expr/context.cc:84]: This is changing the semantics I believe? Previously `has(connection.dns_san_peer_certificate)` would evaluate to false because the attribute is missing. Now it's evaluating to false because it's present but has the value `null`?

**Review by kyessenov — CHANGES_REQUESTED** [review]

File: source/extensions/filters/common/expr/context.cc (hunk 8 above)
Comment (agrawroh) [reply source/extensions/filters/common/expr/context.cc:84]: I agree. This would return an optional containing a `null` CelValue. I'll switch it to `{}` like before. Apologies for the oversight.

**Review by kyessenov — APPROVED** [review]

File: source/extensions/filters/common/expr/context.cc [hunk 9]
```diff
@@ -29,119 +29,455 @@ convertHeaderEntry(Protobuf::Arena& arena,
...
+           }},
+          {Size,
+           [](const RequestWrapper& wrapper) -> absl::optional<CelValue> {
+             if (wrapper.headers_.value_ != nullptr &&
+                 wrapper.headers_.value_->ContentLength() != nullptr) {
+               int64_t length;
+               if (absl::SimpleAtoi(wrapper.headers_.value_->getContentLengthValue(), &length)) {
+                 return CelValue::CreateInt64(length);
+               }
+             }
```
Comment (kyessenov) [inline source/extensions/filters/common/expr/context.cc:126]: This fallthrough slightly changes the implementation - if the value is invalid int, it'll report `bytesReceives`. It probably makes more sense to report something in this case.

**Comment by kyessenov** [comment]
@agrawroh What's up with filter state benchmark:

```
bmFilterState/10000000            3806409 us      3806192 us            1
```

Is it a benchmark problem or some implementation difference?

**Comment by agrawroh** [comment]
> @agrawroh What's up with filter state benchmark:
> 
> ```
> bmFilterState/10000000            3806409 us      3806192 us            1
> ```
> 
> Is it a benchmark problem or some implementation difference?

I think it's just a benchmark issue. I will take a look and see if I can improve it
